source/iterator/pub_sub_listener_test.go

```diff
@@ -118,3 +120,64 @@ func TestNewCDCIterator(t *testing.T) {
 	assert.Equal(t, response.key, res.key)
 	assert.Equal(t, response.psc, res.psc)
 }
+
+func TestNewCDCIterator_Next(t *testing.T) {
+	redisChannel := "subchannel"
+	testMessage := "some_dummy_message"
+	testMessage1 := "some_dummy_message1"
+	mr, err := miniredis.Run()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer mr.Close()
+	conn, err := redis.Dial("tcp", mr.Addr())
+	if err != nil {
+		t.Fatal(err)
+	}
+	response := PubSubIterator{
+		key:     redisChannel,
+		psc:     &redis.PubSubConn{Conn: conn},
+		records: []sdk.Record{},
+		mux:     &sync.Mutex{},
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	res, err := NewPubSubIterator(ctx, conn, redisChannel)
+	assert.Nil(t, err)
+	assert.NotNil(t, res)
+	assert.Equal(t, response.key, res.key)
+	assert.Equal(t, response.psc, res.psc)
+	// publish is a fire and forget method, give a few ms for goroutines to start
+	// otherwise messages might be lost and tests will fail
+	time.Sleep(10 * time.Millisecond)
+	mr.Publish(redisChannel, testMessage)
+	mr.Publish(redisChannel, testMessage1)
+
+	var rec sdk.Record
+	for ctx.Err() == nil {
+		rec, err = res.Next(ctx)
+		if err != nil && err == sdk.ErrBackoffRetry {
+			t.Log("backoff received, waiting for 400ms")
+			time.Sleep(400 * time.Millisecond)
+			continue
+		}
```

Comment:
Is there a way to guarantee this test won't be hanging because of this loop becoming infinite? I do see `ctx` has a timeout, but we don't really have a guarantee that the context will be respected.

If too time consuming, then let's leave a todo for it here (assuming you agree an infinite loop is possible).

Reply:
No, it won't block, We are reading for the data/timeout in the select block, after the timeout the test will exit and go routines will be stopped.

Reply:
You mean the select block in the iterator's Next() method?

Reply:
Yes, once the context is canceled, we will get a context error/tomb dying error, resulting in the loop being stopped.

Reply:
The implementation as you described sounds very reasonable to me. However, the test should be there as a guard against any changes in the implementation, which would, for example, make the loop an infinite one.

Reply:
Refactored the code to use ticker and stop after 10 retries